test/irmin-pack/test_pack.ml

```diff
@@ -280,6 +284,77 @@ module Pack = struct
         Alcotest.fail "Add after closing the index should not be allowed")
       (function I.Closed -> Lwt.return_unit | exn -> Lwt.fail exn)
 
+  (** Index can be flushed to disk independently of pack, we simulate this in
+      the tests using [Index.filter] and [Index.flush]. *)
+  let readonly_sync_index_flush () =
+    Context.get_pack () >>= fun t ->
+    t.clone_index_pack ~readonly:true >>= fun (i, r) ->
+    let ro_sync () =
+      Index.ro_sync i;
+      Pack.ro_sync r
+    in
+    let test w =
+      let x1 = "foo" in
+      let h1 = sha1 x1 in
+      Pack.unsafe_append w h1 x1;
+      ro_sync ();
+      Pack.find r h1 >>= fun y1 ->
+      Alcotest.(check (option string)) "sync before filter" None y1;
+      Index.filter t.index (fun _ -> true);
+      ro_sync ();
+      Lwt.catch
+        (fun () ->
+          Pack.find r h1 >|= fun _ ->
+          Alcotest.fail "index is flushed but not pack, should raise exception")
+        (function Pack.Invalid_read -> Lwt.return_unit | exn -> Lwt.fail exn)
+      >>= fun () ->
+      let x2 = "toto" in
+      let h2 = sha1 x2 in
+      Pack.unsafe_append w h2 x2;
+      Index.flush t.index;
+      ro_sync ();
+      Lwt.catch
+        (fun () ->
+          Pack.find r h2 >|= fun _ ->
+          Alcotest.fail "index is flushed but not pack, should raise exception")
+        (function Pack.Invalid_read -> Lwt.return_unit | exn -> Lwt.fail exn)
+    in
+    test t.pack >>= fun () ->
+    Context.close t.index t.pack >>= fun () -> Context.close i r
+
+  let readonly_find_index_flush () =
+    Context.get_pack () >>= fun t ->
+    t.clone_index_pack ~readonly:true >>= fun (i, r) ->
+    let check h x msg =
+      Pack.find r h >|= fun y -> Alcotest.(check (option string)) msg (Some x) y
+    in
+    let test w =
+      let x1 = "foo" in
+      let h1 = sha1 x1 in
+      Pack.unsafe_append w h1 x1;
+      Pack.sync t.pack;
+      Pack.ro_sync r;
+      check h1 x1 "find before filter" >>= fun () ->
+      Index.filter t.index (fun _ -> true);
+      check h1 x1 "find after filter" >>= fun () ->
+      let x2 = "bar" in
+      let h2 = sha1 x2 in
+      Pack.unsafe_append w h2 x2;
+      Pack.sync t.pack;
+      Pack.ro_sync r;
+      check h2 x2 "find before flush" >>= fun () ->
+      let x3 = "toto" in
+      let h3 = sha1 x3 in
+      Pack.unsafe_append w h3 x3;
+      Index.flush t.index;
+      check h2 x2 "find after flush" >>= fun () ->
+      Pack.sync t.pack;
+      Pack.ro_sync r;
```

Comment:
Once again, I think here the naming of `sync` vs `ro_sync` is misleading / not very descriptive.